From c08579f5416dbb5990c67b6e9c352cad8de26b14 Mon Sep 17 00:00:00 2001
From: Duy Truong <duytruong@google.com>
Date: Mon, 4 May 2026 16:06:15 -0700
Subject: [PATCH] Add bracket checking support to SQLiteTokenizer

Add OPTION_CHECK_BRACKETS to SQLiteTokenizer to allow checking for
unbalanced brackets. This can be used to prevent SQL injection where
user-provided selection strings attempt to break out of parentheses.

Bug: 393377017
Test: atest SQLiteTokenizerTest
Flag: EXEMPT BUGFIX

Autogenerated by MendIt (go/androidmendit)
Cherrypick-From: https://googleplex-android-review.googlesource.com/q/commit:f80be5110a5652541944503569c9e5d69a867c43
Merged-In: I84dd342b7ee2328413c9aef5734d7ffefafa4d85
Change-Id: I84dd342b7ee2328413c9aef5734d7ffefafa4d85
---
 .../database/sqlite/SQLiteTokenizer.java      | 47 ++++++++++++++++---
 .../database/sqlite/SQLiteTokenizerTest.java  | 31 ++++++++++++
 2 files changed, 72 insertions(+), 6 deletions(-)

diff --git a/core/java/android/database/sqlite/SQLiteTokenizer.java b/core/java/android/database/sqlite/SQLiteTokenizer.java
index 7e7c3fb976c70..3ebeb179a5ba3 100644
--- a/core/java/android/database/sqlite/SQLiteTokenizer.java
+++ b/core/java/android/database/sqlite/SQLiteTokenizer.java
@@ -69,6 +69,12 @@ private static char peek(String s, int index) {
      */
     public static final int OPTION_TOKEN_ONLY = 1 << 0;
 
+    /**
+     * Checks brackets; an unbalanced expression or one that attempts to close an
+     * earlier bracket will result in an exception.
+     */
+    public static final int OPTION_CHECK_BRACKETS = 1 << 1;
+
     /**
      * Tokenize the given SQL, returning the list of each encountered token.
      *
@@ -86,11 +92,13 @@ public static List<String> tokenize(@Nullable String sql, int options) {
      *
      * @throws IllegalArgumentException if invalid SQL is encountered.
      */
-    public static void tokenize(@Nullable String sql, int options, Consumer<String> checker) {
+    public static void tokenize(@Nullable String sql, int options,
+            @Nullable Consumer<String> checker) {
         if (sql == null) {
             return;
         }
         int pos = 0;
+        int level = 0;
         final int len = sql.length();
         while (pos < len) {
             final char ch = peek(sql, pos);
@@ -105,7 +113,9 @@ public static void tokenize(@Nullable String sql, int options, Consumer<String>
                 final int end = pos;
 
                 final String token = sql.substring(start, end);
-                checker.accept(token);
+                if (checker != null) {
+                    checker.accept(token);
+                }
 
                 continue;
             }
@@ -142,9 +152,11 @@ public static void tokenize(@Nullable String sql, int options, Consumer<String>
                     } else {
                         token = tokenUnquoted;
                     }
-                    checker.accept(token);
+                    if (checker != null) {
+                        checker.accept(token);
+                    }
                 } else {
-                    if ((options &= OPTION_TOKEN_ONLY) != 0) {
+                    if ((options & OPTION_TOKEN_ONLY) != 0) {
                         throw genException("Non-token detected", sql);
                     }
                 }
@@ -164,10 +176,12 @@ public static void tokenize(@Nullable String sql, int options, Consumer<String>
 
                 final String token = sql.substring(quoteStart + 1, quoteEnd);
 
-                checker.accept(token);
+                if (checker != null) {
+                    checker.accept(token);
+                }
                 continue;
             }
-            if ((options &= OPTION_TOKEN_ONLY) != 0) {
+            if ((options & OPTION_TOKEN_ONLY) != 0) {
                 throw genException("Non-token detected", sql);
             }
 
@@ -199,11 +213,32 @@ public static void tokenize(@Nullable String sql, int options, Consumer<String>
                 throw genException("Semicolon is not allowed", sql);
             }
 
+            if (ch == '(') {
+                pos++;
+                level++;
+
+                continue;
+            }
+
+            if (ch == ')') {
+                pos++;
+                level--;
+                if (level < 0 && (options & OPTION_CHECK_BRACKETS) != 0) {
+                    throw genException("Unbalanced brackets", sql);
+                }
+
+                continue;
+            }
+
             // For this purpose, we can simply ignore other characters.
             // (Note it doesn't handle the X'' literal properly and reports this X as a token,
             // but that should be fine...)
             pos++;
         }
+
+        if (level != 0 && (options & OPTION_CHECK_BRACKETS) != 0) {
+            throw genException("Unbalanced brackets", sql);
+        }
     }
 
     /**
diff --git a/core/tests/coretests/src/android/database/sqlite/SQLiteTokenizerTest.java b/core/tests/coretests/src/android/database/sqlite/SQLiteTokenizerTest.java
index ced18469c6d02..f029c3a4976e6 100644
--- a/core/tests/coretests/src/android/database/sqlite/SQLiteTokenizerTest.java
+++ b/core/tests/coretests/src/android/database/sqlite/SQLiteTokenizerTest.java
@@ -164,10 +164,41 @@ public void testSemicolons() {
         checkTokens("`;`", ";");
     }
 
+    @Test
+    public void testCheckBrackets() {
+        SQLiteTokenizer.tokenize("((a) b)", SQLiteTokenizer.OPTION_CHECK_BRACKETS);
+        try {
+            SQLiteTokenizer.tokenize("((a) b", SQLiteTokenizer.OPTION_CHECK_BRACKETS);
+            fail("Should have thrown");
+        } catch (IllegalArgumentException expected) {
+            assertEquals("Unbalanced brackets in '((a) b'", expected.getMessage());
+        }
+        try {
+            SQLiteTokenizer.tokenize("(a) b)", SQLiteTokenizer.OPTION_CHECK_BRACKETS);
+            fail("Should have thrown");
+        } catch (IllegalArgumentException expected) {
+            assertEquals("Unbalanced brackets in '(a) b)'", expected.getMessage());
+        }
+        try {
+            SQLiteTokenizer.tokenize("1=1)) OR ((1=1", SQLiteTokenizer.OPTION_CHECK_BRACKETS);
+            fail("Should have thrown");
+        } catch (IllegalArgumentException expected) {
+            assertEquals("Unbalanced brackets in '1=1)) OR ((1=1'", expected.getMessage());
+        }
+    }
+
     @Test
     public void testTokens() {
         checkTokens("a,abc,a00b,_1,_123,abcdef", "a abc a00b _1 _123 abcdef");
         checkTokens("a--\nabc/**/a00b''_1'''ABC'''`_123`abc[d]\"e\"f",
                 "a abc a00b _1 _123 abc d e f");
     }
+
+    @Test
+    public void testCheckBracketsInComments() {
+        SQLiteTokenizer.tokenize("((a) -- comment with unmatched bracket )\n b)",
+                SQLiteTokenizer.OPTION_CHECK_BRACKETS);
+        SQLiteTokenizer.tokenize("((a) /* comment with unmatched bracket ) */ b)",
+                SQLiteTokenizer.OPTION_CHECK_BRACKETS);
+    }
 }
